---
title: "Exploratory analysis of a GEO dataset"
author: "Aina Montalban"
date: "5/15/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

what is this exercise is about
```{r}

```

# Summary

This study entitled 'Genome-wide analysis of uterine natural killer cell gene expression from control diet-fed and high fat diet-fed female mice in early pregnancy' was done using mice and expression microarrays.

The following description was given: *Analysis of gene expression in uterine natural killer cells purified from control, diet-induced obese and diet-induced obese resistant female mice. The aim of the study is to identify differentially expressed genes and pathways in uterine natural killer cells affected by maternal diet and obesity in pregnancy.*
  	
## Experimental design

Total RNA isolated from FACS-sorted uterine natural killer cells from Gd10.5 pregnancies in 10 control, 6 diet-induced obese and 4 diet-induced obese resistant mice.

# Citation(s)
Baltayeva J, Beristain A

# Data Description
We have two files:

- The *targets* file.
- The *expression matrix* file.

```{r readTargets}
targets <- read.csv("targets.txt", sep = "")
expressions <- read.delim("expressions.txt", row.names = 1)
```

```{r}
show(targets)

```


```{r}
dim(expressions)
summary(expressions)
```

```{r}
boxplot(expressions, col = targets$color)
```


## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
