---
title: "Exploratory analysis of a GEO dataset"
author: "Aina Montalban"
date: "5/15/2019"
output:
  pdf_document: default
  html_document: default
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

The aim of this exercise is to practice the basic elements of Bioconductor and Rmarkdown using a study from the Gene Expression Omnibus Database.
```{r}

```

# Summary

This study entitled 'Genome-wide analysis of uterine natural killer cell gene expression from control diet-fed and high fat diet-fed female mice in early pregnancy' was done using mice and expression microarrays.

The following description was given: *Analysis of gene expression in uterine natural killer cells purified from control, diet-induced obese and diet-induced obese resistant female mice. The aim of the study is to identify differentially expressed genes and pathways in uterine natural killer cells affected by maternal diet and obesity in pregnancy.*
  	
## Experimental design

Total RNA isolated from FACS-sorted uterine natural killer cells from Gd10.5 pregnancies in 10 control, 6 diet-induced obese and 4 diet-induced obese resistant mice.

# Citation(s)
Baltayeva J, Beristain A

# Study Description
We have two files:

- The *targets* file.
- The *expression matrix* file.

# Read or create the data 
The ExpressionSet object can be created in three ways: (i) using R objects created from the files downloaded (ii) from scratch, creating and collating the different components, and (iii) from the web using geoQUERY. 

```{r readTargets}
targets <- read.csv("targets.txt", sep = "")
expressions <- as.matrix(read.delim("expressions.txt", row.names = 1))
class(expressions)
```

The package Biobase contains basic structures for microarray data.

```{r}
require(Biobase)
```

```{r}
myEset <- ExpressionSet(expressions)
class(myEset)
show(myEset)
```

The class AnnotatedDataFrame is intended to contain covariate information.
```{r}
columnDesc <-  data.frame(labelDescription= c("Sample Names", "Diet type", "Replicate"))
myAnnotDF <- new("AnnotatedDataFrame", data=targets, varMetadata= columnDesc)
show(myAnnotDF)
phenoData(myEset) <- myAnnotDF
```

The class MIAME was created to add the "Minimum Information About a Microarray Experiment"
```{r}
myDesc <- new("MIAME", name= "Expression profiling by array",
            lab= "Beristain lab",
            contact= "Alexander G Beristain",
            title="Genome-wide analysis of uterine natural killer cell gene expression from control diet-fed and high fat diet-fed female mice in early pregnancy", url="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi")
print(myDesc)
```


```{r}
rownames(pData(myAnnotDF))<-pData(myAnnotDF)$sampleNames
```


```{r}

myEset <- ExpressionSet(assayData=expressions, 
                        phenoData=myAnnotDF,
                        experimentData = myDesc, annotation =  	"GPL23038")
show(myEset)
```

## Access Slot values


```{r}
smallEset <- myEset[1:15,c(1:3,6:8)]
dim(exprs(smallEset))
dim(pData(smallEset))
head(pData(smallEset))
all(colnames(exprs(smallEset))==rownames(pData(smallEset)))


## ------------------------------------------------------------------------
youngEset <- myEset[,pData(myEset)$DietType == "Control_diet_fed"]
dim(exprs(youngEset))
head(pData(youngEset))
```


## The GEOquery package

```{r}
require(GEOquery)
```

```{r}
gse <- getGEO("GSE124927")
class(gse)
names(gse)
gse[[1]]
```

## Data Description
```{r}
dim(expressions)
summary(expressions)
```

## Including Plots
```{r}
boxplot(expressions)
```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
    