---
title: "Exploratory analysis of a GEO dataset"
author: "Aina Montalban"
date: "5/15/2019"
output:
  html_document: default
  pdf_document: default
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

The aim of this exercise is to practice the basic elements of Bioconductor and Rmarkdown using a study from the Gene Expression Omnibus Database.

## Summary

This study entitled 'Genome-wide analysis of uterine natural killer cell gene expression from control diet-fed and high fat diet-fed female mice in early pregnancy' was done using mice and expression microarrays.

The following description was given: *Analysis of gene expression in uterine natural killer cells purified from control, diet-induced obese and diet-induced obese resistant female mice. The aim of the study is to identify differentially expressed genes and pathways in uterine natural killer cells affected by maternal diet and obesity in pregnancy.*
  	
## Experimental design

Total RNA isolated from FACS-sorted uterine natural killer cells from Gd10.5 pregnancies in 10 control, 6 diet-induced obese and 4 diet-induced obese resistant mice.

## Citation(s)
Baltayeva J, Beristain A

## Study Description
We have two files:

- The *targets* file.
- The *expression matrix* file.

# Read the data 
The ExpressionSet object can be created in two ways: (i) using R objects created from the files downloaded and (ii) from the web using geoQUERY. 

```{r readTargets}
targets <- read.csv("targets.txt", sep = "")
expressions <- as.matrix(read.delim("expressions.txt", row.names = 1))
class(expressions)
```


The package Biobase contains basic structures for microarray data.

```{r message=FALSE}
require(Biobase)
```

In the following steps, we will create an ExpressionSet using the GEO study mentioned. 

### Slot AssayData
The main element, and indeed the only one to be provided to create an ExpressionSet is AssayData. It can be seen as a matrix with as many rows as genes or generically “features” and as many columns as samples or individuals.

```{r}
myEset <- ExpressionSet(expressions)
class(myEset)
show(myEset)
```

### Information about covariates
The class AnnotatedDataFrame is intended to contain covariate information.
```{r}
columnDesc <-  data.frame(labelDescription= c("Sample Names", "Group", "Replicate"))
myAnnotDF <- new("AnnotatedDataFrame", data=targets, varMetadata= columnDesc)
show(myAnnotDF)
# We can add it to to the ExpressionSet.
phenoData(myEset) <- myAnnotDF
```

```{r}
rownames(pData(myAnnotDF))<-pData(myAnnotDF)$sampleName
myEset <- ExpressionSet(assayData=expressions, phenoData=myAnnotDF)
show(myEset)
```

## Adding information about features

The number of rows in featureData must match the number of rows in assayData. Row names of featureData must match row names of the matrix in assayData. ALternatively we can simple store the names of the features using a character vector in the slot featureNames.

```{r}

myGenes <- rownames(expressions)
myEset <- ExpressionSet(assayData=expressions, phenoData=myAnnotDF, featureNames= myGenes)
show(myEset)
```


### Information about the experiment
The class MIAME was created to add the "Minimum Information About a Microarray Experiment"
```{r}
myDesc <- new("MIAME", name= "Expression profiling by array",
            lab= "Beristain lab",
            contact= "Alexander G Beristain",
            title="Genome-wide analysis of uterine natural killer cell gene expression from control diet-fed and high fat diet-fed female mice in early pregnancy", url="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi")
print(myDesc)
```

We can also add the annotation, which is the platform. 
```{r}
myEset <- ExpressionSet(assayData=expressions, 
                        phenoData=myAnnotDF,
                        experimentData = myDesc,featureNames = myGenes, annotation =  	"GPL23038")
show(myEset)
```
# Using objects of class ExpressionSet

## Access Slot values
```{r}
dim(exprs(myEset))
class(phenoData(myEset))
class(pData(phenoData(myEset)))
head(pData(phenoData(myEset)))
head(pData(myEset))
```

## Subsetting
```{r}
smallEset <- myEset[1:15,c(1:3,6:8)]
dim(exprs(smallEset))
dim(pData(smallEset))
head(pData(smallEset))
all(colnames(exprs(smallEset))==rownames(pData(smallEset)))
youngEset <- myEset[,pData(myEset)$DietType == "Control_diet_fed"]
dim(exprs(youngEset))
```

## The GEOquery package
The GEOquery package allows us to to get data from GEO very easily. There is only one command that is need, getGEO.
```{r message=FALSE}
require(GEOquery)
```

```{r warning=FALSE}
gse <- getGEO("GSE124927")
class(gse)
names(gse)
```

```{r}
gse[[1]]
```


# Data Description

Moreover, we can describe the data we have by  doing plots or other commands.

## Dimension and Summary
```{r}
dim(expressions)
summary(expressions)
```

## Including Plots
```{r}
boxplot(expressions)
```


    